src/java.base/share/classes/java/lang/foreign/Linker.java

```diff
@@ -366,10 +366,18 @@
  * <h3 id="variadic-funcs">Variadic functions</h3>
  *
  * Variadic functions (e.g. a C function declared with a trailing ellipses {@code ...} at the end of the formal parameter
- * list or with an empty formal parameter list) are not supported directly by the native linker. However, it is still possible
- * to link a variadic function by using a <em>specialized</em> function descriptor, together with a
- * {@linkplain Linker.Option#firstVariadicArg(int) a linker option} which indicates the position of the first variadic argument
- * in that specialized descriptor.
+ * list or with an empty formal parameter list, the latter also being called <em>prototype-less</em>) are not supported
+ * directly by the native linker. However, it is still possible to link a variadic function by using a <em>specialized</em>
+ * function descriptor, together with {@linkplain Linker.Option#firstVariadicArg(int) a linker option} which indicates the
+ * position of the first variadic argument in that specialized descriptor. The corresponding argument layout, and all following
+ * argument layouts in the descriptor, are considered to be <em>variadic argument layouts</em>. These layouts correspond
+ * to the arguments passed in place of the ellipsis {@code ...}, or to the arguments passed to a prototype-less function.
+ * <p>
+ * It should be noted that values passed as variadic arguments undergo default argument promotion in C. Each value of
+ * type {@code float} is converted to {@code double}, and each integral type smaller than {@code int} is converted to
+ * {@code int}. As such, the native linker will reject attempts to link function descriptors with certain variadic argument
+ * layouts. Namely, {@linkplain ValueLayout value layouts} that have a carrier type of {@code boolean}, {@code byte},
```

Comment:
Is there any reason as to why you decided to say which layouts are **not** allowed as variadic layouts? I'm wondering whether, if we add more carriers in the future, this list will probably need to be updated? Or do the restriction only involve these "small" types, and stuff like `long double` is always allowed?

Reply:
No real reason for the current polarity. The specification explicitly calls out the float -> double conversion, and then for the integral types it refers to 'integer promotions' [1], which is a somewhat complex rule set for assigning ranks to integer types.

So, I think if we add more carriers in the future, we would have to worry about integral types that have a rank less than (unsigned) int when considering this list. This is currently only the types I've listed (though, `char` is a bit of a special case), and this seems relatively stable to me. e.g. `long double` would not need to be added here.

[1]: [Integer_promotions](https://en.cppreference.com/w/c/language/conversion#Integer_promotions)